File: code-tiny/tools/spring/annotation_catalog.py
from __future__ import annotations

from typing import Dict, FrozenSet, Iterable, Optional


def short_annotation_name(name: str) -> str:
    raw = (name or "").strip()
    if ":" in raw:
        raw = raw.split(":", 1)[1]
    return raw.rsplit(".", 1)[-1]


APPLICATION_ANNOTATIONS: FrozenSet[str] = frozenset({"SpringBootApplication", "EnableAutoConfiguration"})
CONFIGURATION_ANNOTATIONS: FrozenSet[str] = frozenset({"Configuration"})
BEAN_ANNOTATIONS: FrozenSet[str] = frozenset({"Bean"})
COMPONENT_ANNOTATIONS: FrozenSet[str] = frozenset({"Component", "Service", "Repository", "Controller", "RestController"})
CONTROLLER_ANNOTATIONS: FrozenSet[str] = frozenset({"Controller", "RestController"})
SERVICE_ANNOTATIONS: FrozenSet[str] = frozenset({"Service"})
REPOSITORY_ANNOTATIONS: FrozenSet[str] = frozenset({"Repository"})
INJECTION_ANNOTATIONS: FrozenSet[str] = frozenset({"Autowired", "Inject", "Resource"})
VALUE_ANNOTATIONS: FrozenSet[str] = frozenset({"Value", "ConfigurationProperties"})
HTTP_MAPPING_METHODS: Dict[str, str] = {
    "GetMapping": "GET",
    "PostMapping": "POST",
    "PutMapping": "PUT",
    "PatchMapping": "PATCH",
    "DeleteMapping": "DELETE",
}
HTTP_MAPPING_ANNOTATIONS: FrozenSet[str] = frozenset({"RequestMapping", *HTTP_MAPPING_METHODS.keys()})
ENTITY_ANNOTATIONS: FrozenSet[str] = frozenset({"Entity", "MappedSuperclass", "Embeddable"})
REPOSITORY_SUPERTYPES: FrozenSet[str] = frozenset({
    "Repository",
    "CrudRepository",
    "JpaRepository",
    "PagingAndSortingRepository",
    "ReactiveCrudRepository",
})
TRANSACTION_ANNOTATIONS: FrozenSet[str] = frozenset({"Transactional"})
MESSAGE_LISTENER_ANNOTATIONS: FrozenSet[str] = frozenset({"KafkaListener", "RabbitListener", "RabbitHandler"})
SCHEDULED_ANNOTATIONS: FrozenSet[str] = frozenset({"Scheduled"})
ASYNC_ANNOTATIONS: FrozenSet[str] = frozenset({"Async"})
EVENT_LISTENER_ANNOTATIONS: FrozenSet[str] = frozenset({"EventListener", "TransactionalEventListener"})
SECURITY_METHOD_ANNOTATIONS: FrozenSet[str] = frozenset({
    "PreAuthorize",
    "PostAuthorize",
    "PreFilter",
    "PostFilter",
    "Secured",
    "RolesAllowed",
})
AOP_CLASS_ANNOTATIONS: FrozenSet[str] = frozenset({"Aspect"})
AOP_METHOD_ANNOTATIONS: FrozenSet[str] = frozenset({
    "Pointcut",
    "Before",
    "After",
    "AfterReturning",
    "AfterThrowing",
    "Around",
})
VALIDATION_ANNOTATIONS: FrozenSet[str] = frozenset({
    "Valid",
    "Validated",
    "NotNull",
    "NotBlank",
    "NotEmpty",
    "Size",
    "Min",
    "Max",
    "Pattern",
    "Email",
    "Positive",
    "PositiveOrZero",
    "Negative",
    "NegativeOrZero",
})
CACHE_ANNOTATIONS: FrozenSet[str] = frozenset({"Cacheable", "CachePut", "CacheEvict", "Caching"})


def has_any_annotation(names: Iterable[str], catalog: Iterable[str]) -> bool:
    known = set(catalog)
    return any(short_annotation_name(item) in known for item in names)


def first_annotation(names: Iterable[str], catalog: Iterable[str]) -> Optional[str]:
    known = set(catalog)
    for item in names:
        short = short_annotation_name(item)
        if short in known:
            return short
    return None

File: rust/crates/analyzer-jvm-overlays/src/spring/annotation_catalog.rs
//! Port `tools/spring/annotation_catalog.py`.

use std::collections::{BTreeMap, BTreeSet};

pub fn short_annotation_name(name: &str) -> String {
    let mut raw = name.trim().to_string();
    if let Some(index) = raw.find(':') {
        raw = raw[index + 1..].to_string();
    }
    match raw.rfind('.') {
        Some(index) => raw[index + 1..].to_string(),
        None => raw,
    }
}

pub fn application_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["SpringBootApplication", "EnableAutoConfiguration"])
}
pub fn configuration_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Configuration"])
}
pub fn bean_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Bean"])
}
pub fn component_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Component", "Service", "Repository", "Controller", "RestController"])
}
pub fn controller_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Controller", "RestController"])
}
pub fn service_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Service"])
}
pub fn repository_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Repository"])
}
pub fn injection_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Autowired", "Inject", "Resource"])
}
pub fn value_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Value", "ConfigurationProperties"])
}
pub fn http_mapping_methods() -> BTreeMap<&'static str, &'static str> {
    BTreeMap::from([
        ("GetMapping", "GET"),
        ("PostMapping", "POST"),
        ("PutMapping", "PUT"),
        ("PatchMapping", "PATCH"),
        ("DeleteMapping", "DELETE"),
    ])
}
pub fn http_mapping_annotations() -> BTreeSet<&'static str> {
    let mut set: BTreeSet<&'static str> = BTreeSet::from(["RequestMapping"]);
    set.extend(http_mapping_methods().keys().copied());
    set
}
pub fn entity_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Entity", "MappedSuperclass", "Embeddable"])
}
pub fn repository_supertypes() -> BTreeSet<&'static str> {
    BTreeSet::from([
        "Repository",
        "CrudRepository",
        "JpaRepository",
        "PagingAndSortingRepository",
        "ReactiveCrudRepository",
    ])
}
pub fn transaction_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Transactional"])
}
pub fn message_listener_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["KafkaListener", "RabbitListener", "RabbitHandler"])
}
pub fn scheduled_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Scheduled"])
}
pub fn async_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Async"])
}
pub fn event_listener_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["EventListener", "TransactionalEventListener"])
}
pub fn security_method_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["PreAuthorize", "PostAuthorize", "PreFilter", "PostFilter", "Secured", "RolesAllowed"])
}
pub fn aop_class_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Aspect"])
}
pub fn aop_method_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Pointcut", "Before", "After", "AfterReturning", "AfterThrowing", "Around"])
}
pub fn validation_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from([
        "Valid",
        "Validated",
        "NotNull",
        "NotBlank",
        "NotEmpty",
        "Size",
        "Min",
        "Max",
        "Pattern",
        "Email",
        "Positive",
        "PositiveOrZero",
        "Negative",
        "NegativeOrZero",
    ])
}
pub fn cache_annotations() -> BTreeSet<&'static str> {
    BTreeSet::from(["Cacheable", "CachePut", "CacheEvict", "Caching"])
}
